// n8n-nodes-grok/dist/credentials/GrokApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrokApi = void 0;
class GrokApi {
    constructor() {
        this.name = 'grokApi';
        this.displayName = 'Grok API';
        this.documentationUrl = 'https://console.x.ai/';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                required: true,
                description: 'The Grok API key',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'Authorization': '=Bearer {{$credentials.apiKey}}',
                    accept: 'application/json',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.x.ai/v1',
                url: '/models',
            },
        };
    }
}
exports.GrokApi = GrokApi;
